source-dropbox/main.go

```diff
@@ -0,0 +1,308 @@
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	oauthClient := cfg.Credentials.getOauthClient(ctx)
+	config := dropbox.Config{
+		Client: oauthClient,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks if the specified path exists in the Dropbox store.
+// It returns an error if the path does not exist or if there was an error
+// while checking the path.
+func (dbStore *dropboxStore) check() error {
+	_, err := dbStore.client.ListFolder(&files.ListFolderArg{Path: dbStore.config.Path, Limit: 1})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (dbStore *dropboxStore) List(ctx context.Context, query filesource.Query) (filesource.Listing, error) {
+	// The Dropbox API returns pages of entries sorted by their modification date and time. However for
+	// the connector to work with the `filesource` package correctly, the listing returned by `List` must
+	// contain all entries sorted in lexicographic, monotonic order. To achieve this, we paginate through
+	// all entries, convert them to `filesource.ObjectInfo`s, and sort them based on their path.
+
+	// Get the first page of entries at the specified path from Dropbox.
+	arg := &files.ListFolderArg{Path: query.Prefix, Recursive: query.Recursive}
+	f, err := dbStore.client.ListFolder(arg)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving folder: %w", err)
+	}
+
+	entries := f.Entries
+
+	// Get the remaining pages of entries.
+	for f.HasMore {
+		arg := files.NewListFolderContinueArg(f.Cursor)
+		f, err = dbStore.client.ListFolderContinue(arg)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving additional files: %w", err)
+		}
+
+		entries = append(entries, f.Entries...)
+	}
+
+	// Convert each file and folder entry to an ObjectInfo.
+	var objects []filesource.ObjectInfo
+	for _, entry := range entries {
+		var obj filesource.ObjectInfo
+
+		switch e := entry.(type) {
+		case *files.FileMetadata:
+			obj = filesource.ObjectInfo{
+				Path:       e.PathLower,
+				IsPrefix:   false,
+				ContentSum: e.ContentHash,
+				Size:       int64(e.Size),
+				ModTime:    e.ServerModified,
+			}
+
+		case *files.FolderMetadata:
+			if !query.Recursive {
+				obj = filesource.ObjectInfo{
+					Path:     e.PathLower,
+					IsPrefix: true,
+				}
+			}
+
+		default:
+			// Ignore all other metadata types (like DeletedMetadata).
+			continue
+		}
+
+		// Only add files and folders that are lexicographically 
+		// greater than or equal to the query's StartAt property.
+		if query.StartAt < obj.Path {
+			objects = append(objects, obj)
```

Comment:
What happens in the case of `*files.FolderMetadata` and when `query.Recursive` is `true`? I guess `obj.Path` would be `""` and so the empty object wouldn't be appended?

Reply:
That's what I'm thinking & seeing when I test it out. But that's not super clear, so what do you think about re-writing the `*files.FolderMetadata` case like below?
```golang
case *files.FolderMetadata:
	if query.Recursive {
		continue
	}
	obj = filesource.ObjectInfo{
		Path:     e.PathLower,
		IsPrefix: true,
	}
```

Then we wouldn't reach `if query.StartAt < obj.Path` when `query.Recursive` is `true`.

Reply:
That seems better!

Reply:
Updated!